pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1172,29 +1198,88 @@ func (vs *VSphere) CreateVolume(volumeOptions *vclib.VolumeOptions) (canonicalVo
 				return "", err
 			}
 		} else {
-			// Since no storage policy is specified but datastore is specified, check
-			// if the given datastore is a shared datastore across all node VMs.
-			sharedDsList, err := getSharedDatastoresInK8SCluster(ctx, dc, vs.nodeManager)
-			if err != nil {
-				klog.Errorf("Failed to get shared datastore: %+v", err)
-				return "", err
-			}
-			found := false
-			for _, sharedDs := range sharedDsList {
-				if datastore == sharedDs.Info.Name {
-					found = true
-					break
+			// If zone is specified, pick the datastore in the zone with maximum free space within the zone.
+			if volumeOptions.Datastore == "" && len(volumeOptions.Zone) != 0 {
+				klog.V(4).Infof("Specified zone : %s", volumeOptions.Zone)
+				dsList, err = getDatastoresForZone(ctx, dc, vs.nodeManager, volumeOptions.Zone)
+
+				// If unable to get any datastore, fail the operation
+				if len(dsList) == 0 {
+					err := fmt.Errorf("Failed to find a shared datastore matching zone %s", volumeOptions.Zone)
+					klog.Error(err)
+					return "", err
+				}
+
+				if err != nil {
+					return "", err
+				}
+				datastore, err = getMostFreeDatastoreName(ctx, nil, dsList)
+				if err != nil {
+					klog.Errorf("Failed to get shared datastore: %+v", err)
+					return "", err
+				}
+				klog.V(1).Infof("Specified zone : %s. Selected datastore : %s", volumeOptions.StoragePolicyName, datastore)
+			} else {
+				var sharedDsList []*vclib.DatastoreInfo
+				var err error
+				if len(volumeOptions.Zone) == 0 {
+					// If zone is not provided, get the shared datastore across all node VMs.
+					klog.V(4).Infof("Validating if datastore %s is shared across all node VMs", datastore)
+					sharedDsList, err = getSharedDatastoresInK8SCluster(ctx, dc, vs.nodeManager)
+					if err != nil {
+						klog.Errorf("Failed to get shared datastore: %+v", err)
+						return "", err
+					}
+					// Prepare error msg to be used later, if required.
+					err = fmt.Errorf("The specified datastore %s is not a shared datastore across node VMs", datastore)
+				} else {
+					// If zone is provided, get the shared datastores in that zone.
+					klog.V(4).Infof("Validating if datastore %s is in zone %s ", datastore, volumeOptions.Zone)
+					sharedDsList, err = getDatastoresForZone(ctx, dc, vs.nodeManager, volumeOptions.Zone)
+					if err != nil {
+						return "", err
+					}
+					// Prepare error msg to be used later, if required.
+					err = fmt.Errorf("The specified datastore %s does not match the provided zones : %s", datastore, volumeOptions.Zone)
+				}
+				found := false
+				// Check if the selected datastore belongs to the list of shared datastores computed.
+				for _, sharedDs := range sharedDsList {
+					if datastore == sharedDs.Info.Name {
+						klog.V(4).Infof("Datastore validation succeeded")
+						found = true
+						break
+					}
+				}
+				if !found {
+					klog.Error(err)
+					return "", err
```

Comment:
How exhaustive is our e2es to cover this? I think we ought to add some more tests unit/intgeration/e2e that exercise this code path before this change can be merged.

Reply:
Yes. All the cases handled in the above section are code are covered under e2e tests. The PR for tests is this: https://github.com/kubernetes/kubernetes/pull/72732